crates/e2e/src/prover/block.rs

```diff
@@ -1,10 +1,489 @@
+            panic!("Recursive groth16 proofs are not supported in mock mode");
+        }
+        // synchronous mode (cuda, cpu)
+        else {
+            synchronous_prover(start_height, &mut trusted_height, num_blocks, trusted_root).await?
+        }
+    };
+    Ok(proof)
+}
+
+pub async fn parallel_prover(
+    start_height: u64,
+    trusted_height: &mut u64,
+    num_blocks: u64,
+    trusted_root: &mut FixedBytes<32>,
+) -> Result<SP1ProofWithPublicValues, Box<dyn Error>> {
+    let storage_path = dirs::home_dir()
+        .expect("cannot find home directory")
+        .join(".ev-prover")
+        .join("data")
+        .join("proofs.db");
+    let proof_storage = Arc::new(RocksDbProofStorage::new(storage_path)?);
+
+    let genesis_path = dirs::home_dir()
+        .expect("cannot find home directory")
+        .join(".ev-prover")
+        .join("config")
+        .join("genesis.json");
+    let (genesis, chain_spec) = load_chain_spec_from_genesis(genesis_path.to_str().unwrap())?;
+
+    let namespace_hex = env::var("CELESTIA_NAMESPACE").expect("CELESTIA_NAMESPACE must be set");
+    let namespace = Namespace::new_v0(&hex::decode(namespace_hex)?)?;
+
+    let celestia_client = Client::new(config::CELESTIA_RPC_URL, None)
+        .await
+        .context("Failed creating Celestia RPC client")?;
+    let pub_key = get_sequencer_pubkey().await?;
+
+    let block_prover_elf = fs::read("elfs/ev-exec-elf").expect("Failed to read ELF");
+    let client = ProverClient::from_env();
+    let (pk, vk) = client.setup(&block_prover_elf);
+
+    let mut trusted_heights = Vec::new();
+    let mut trusted_roots = Vec::new();
+    trusted_heights.push(*trusted_height);
+    trusted_roots.push(*trusted_root);
+
+    let mut current_trusted_height = *trusted_height;
+    let mut current_trusted_root = *trusted_root;
+
+    // before we generate proofs in parallel mode, we execute all blocks to
+    // collect the trusted height and root to then supply them optimistically to the prover
+    for block_number in start_height..(start_height + num_blocks) {
+        println!("\nProcessing block: {block_number}");
+        let blobs: Vec<Blob> = celestia_client
+            .blob_get_all(block_number, &[namespace])
+            .await
+            .expect("Failed to get blobs")
+            .unwrap_or_default();
+        println!("Got {} blobs for block: {}", blobs.len(), block_number);
+
+        let extended_header = celestia_client
+            .header_get_by_height(block_number)
+            .await
+            .expect("Failed to get extended header");
+        let namespace_data = celestia_client
+            .share_get_namespace_data(&extended_header, namespace)
+            .await
+            .expect("Failed to get namespace data");
+        let mut proofs: Vec<NamespaceProof> = Vec::new();
+        for row in namespace_data.rows {
+            proofs.push(row.proof);
+        }
+        println!("Got NamespaceProofs, total: {}", proofs.len());
```

Comment:
nit; could we add a little bit of "breathing space" for the code here, becomes easier to read when there's a few line breaks imo

```suggestion
        println!("\nProcessing block: {block_number}");
        
        let blobs: Vec<Blob> = celestia_client
            .blob_get_all(block_number, &[namespace])
            .await
            .expect("Failed to get blobs")
            .unwrap_or_default();
        
        println!("Got {} blobs for block: {}", blobs.len(), block_number);

        let extended_header = celestia_client
            .header_get_by_height(block_number)
            .await
            .expect("Failed to get extended header");
        
        let namespace_data = celestia_client
            .share_get_namespace_data(&extended_header, namespace)
            .await
            .expect("Failed to get namespace data");
        
        let mut proofs: Vec<NamespaceProof> = Vec::new();
        for row in namespace_data.rows {
            proofs.push(row.proof);
        }
        
        println!("Got NamespaceProofs, total: {}", proofs.len());
```